### PR TITLE
Implement delivery system

### PR DESCRIPTION
**Pull request by ak47na**
The PR implements the methods in `DeliverySystem.java`:

When a courier makes a delivery slot request, `processDeliveryRequest(DeliverySlot deliverySlot)` inside `JourneyHandler` will call `deliverySystem.findOptimalDeliveryJourneyForDeliverySlot(deliverySlot)`.
The method will first assign a set of orders that fit the requirements of deliverySlot by calling `findOrdersForDeliveryRequest`. Then,  it will return an `ArrayList` with the waypoints of the journey, starting from the last one that must be visited.

### CODE REVIEW

**Review by anamariamilcu — APPROVED** [review]

File: src/main/java/com/google/sps/data/Journey.java [hunk 1]
```diff
@@ -56,12 +55,59 @@ public Pair(int first, int second) {
...
     restrictions = new ArrayList<>();
     minTime = 0;
   }
+
+  public int getNumberOfWaypoints() {
+    throw new UnsupportedOperationException("TODO: Implement this method.");
+  }
+
+  /**
+   * Using the bestTime matrix computed in findOptimalTimePaths(), it returns true if the minimum
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: `findOptimalTimePaths` does not exist. `bestTime` does not exist.

File: src/test/java/com/google/sps/data/MapsRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Google LLC
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/MapsRequestTest.java]: This file is unrelated and should be it's own PR as it's adding tests to a class added in #6

File: src/main/java/com/google/sps/data/DeliverySystem.java [hunk 3]
```diff
@@ -26,16 +26,74 @@
...
+        (int) order.getProperty(OrderHandler.OrderProperty.LIBRARY_ID.label));
+      Point recipient = new Point((double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LAT.label), 
+        (double) order.getProperty(OrderHandler.OrderProperty.RECIPIENT_LNG.label));
+      
+      // Check if the library is already a waypoint in the journey, and add it in case it's not.
+      boolean libraryIsWaypoint = !journey.addPointToWaypoints(library);
+      // Check if the recipient is already a waypoint in the journey, and add it in case it's not.
+      boolean recipientIsWaypoint = !journey.addPointToWaypoints(recipient);
+
+      if (journey.getNumberOfWaypoints() > MAX_WAYPOINTS || !journey.findJourneyForTimeslot(deliverySlot)) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/DeliverySystem.java:67]: Calling findJourneyForTimeslot after you add every slot requires rerunning your optimization algorithm from scratch. And the logic here for handling max waypoints seems super complicated.

Why not just add 25 waypoints, optimize the route and the trim the waypoints that exceed your delivery slot.

File: src/main/java/com/google/sps/data/DeliverySystem.java [hunk 4]
```diff
@@ -26,16 +26,74 @@
...
+          // delivery slot.
+          journey.removeWaypoint(recipient);
+        }
+      } else {
+        // The library must be visited before the recipient.
+        journey.addRestriction(journey.getWaypointIndex(library), journey.getWaypointIndex(recipient));
+        // Add the books from current order to the library their are assigned to.
+        library.addBooks((List<String>)order.getProperty("books"));
+        // The order is assigned to the delivery slot and its status changes in datastore.
+        order.setProperty("status", "ASSIGNED");
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/DeliverySystem.java]: You really don't want to do perform database mutations in a method called `find`. That will confuse everyone. It's also a really bad idea to tie your algorithm logic with your database mutations.

Do this instead:
- Run your algorithm, pick your waypoints, do all your sanity checks and very that everything is fine, whatever it is you need to do. <= ideally this class would only do this.
- Then commit your mutations to the database: create the journey for the courier in the database, mark the orders as assigned to the journey, etc.

Ideally the commit part should and in a transaction (https://cloud.google.com/datastore/docs/concepts/transactions) to prevent an error from actually committing invalid state to the database (e.g. the order is marked as assigned, but noone is actually delivery it). But a TODO is totally fine for now.

Having your algorithm perform database mutations "as it goes" has several disadvantages:
- Hard to test. Your tests not only need to test the final output, but also database mutations (you don't seem to be doing that in your tests, btw)
- Hard to move to use a transaction: mutations happen all over the place in many classes. Even if you do manage to make them work, your transactions would be horribly inefficient. You generally want transactions to be quick, without having to have a transaction live for the life of a potentially slow running algorithm.
- Hard to refactor: I want to change the database, now I need to have a deep understanding of the routing algorithm... I want to change the algorithm, I need to understand datastore schema... why?

File: src/main/java/com/google/sps/data/DeliverySystem.java [hunk 5]
```diff
@@ -26,16 +26,74 @@
...
+  public ArrayList<Point> findOptimalDeliveryJourneyForDeliverySlot(DeliverySlot deliverySlot) throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
+    Journey journey = findOrdersForDeliveryRequest(deliverySlot);
+    return journey.findOptimalOrderForWaypoints();
+  }
+
+  /**
+   * Assigns a set of orders for the delivery request such that they fit the [startTime, endTime]
+   * timeslot.
+   */
+  public Journey findOrdersForDeliveryRequest(DeliverySlot deliverySlot)  throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/DeliverySystem.java]: nit: This method is called findOrders, but it actually constructs a journey, this is confusing. Call it `createJourneyForDeliverySlot` if that's what it does.

File: src/main/java/com/google/sps/data/DeliverySystem.java [hunk 6]
```diff
@@ -26,16 +26,74 @@
...
+    this.pathFinder = pathFinder;
+  }
+
   /**
    * Finds the order in which points must be visited such that the journey lasts minimum time and it
    * delivers all orders assigned in findOrdersForDeliveryRequest.
    */
-  public ArrayList<Point> findOptimalDeliveryJourneyForDeliverySlot(DeliverySlot deliverySlot) throws ApiException, IOException, InterruptedException {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public ArrayList<Point> findOptimalDeliveryJourneyForDeliverySlot(DeliverySlot deliverySlot) throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/DeliverySystem.java]: You should always try to minimize the number of public methods. After this PR, this public method does not provide value over findOrdersForDeliveryRequest (other than saving you to type a few characters, which is not a good reason).

`findOrdersForDeliveryRequest` is also dead-on-arrival method (only called in tests), which is an antipattern.

You can split them if you find a good splitting point (one method creates a journey, the other takes the journey you just created and "optimizes" it). I would probably just keep one of them though which does everything.

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/DeliverySystem.java]: You'll probably need to store a lot more data here, for each waypoint you need to store what order ids to pick up or deliver, so you can actually display some that data to the courier.

**Review by alexcepoi — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/data/Journey.java (hunk 1 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: findOptimalTimePaths is a private method and bestTime is a local variable. I changed the comment.

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: Yes, for each waypoint, I store a private list with the IDs of the books that must be borrowed/delivered.

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 3 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java:67]: The problem with adding the points first is that libraries and the addresses of recipients are different, e.g. trimming waypoints to fit the delivery slot requirements might result in a set of waypoints containing only libraries.
Another example would be if:
recipient1 orders from library1 and library2
recipient2 orders from library3
The trimmed set can end up with recipient2, library1, library2, which is not valid.

Comment (alexcepoi) [reply src/main/java/com/google/sps/data/DeliverySystem.java:67]: Yeah, after trimming waypoints, you would need to remove any library waypoints from which the courier suddenly has no books to pick up for (if you removed a recipient1, there's nothing to pick up from library1 and library2).

But yeah, you're right that it would then be theoretically possible that a user who lives near many libraries would not get any deliveries (or only library waypoints which would then get removed) if they have a very small timeslot. Like with most algorithms, there are tradeoffs :)

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 6 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: I removed `findOptimalDeliveryJourneyForDeliverySlot` and renamed `findOrdersForDeliveryRequest` to `createJourneyForDeliverySlot`.

File: src/test/java/com/google/sps/data/MapsRequestTest.java (hunk 2 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/MapsRequestTest.java]: Done.

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: I understand, I changed the class so that the status of orders is updated in OrderHandler.java. In the next PR I will use transactions for updates and remove the database logic from DeliverySystem.

File: src/main/java/com/google/sps/data/Journey.java [hunk 7]
```diff
@@ -16,31 +16,32 @@
...
-  // a new Point is added to the hashmap, its value will be the equal to the number of elements.
-  // The hashmap is preferred to an ArrayList so that Point objects can be removed efficiently.
-  private HashMap<Point, Integer> waypoints;
-  // TODO[ak47na]: add PathFinder object, interface and the classes that implement it: 
-  // GoogleMapsPathFinder used for production and ManhattanDistancePathFinder for testing.
+  // Array of Point objects representing the waypoints that must be visited not including the start
+  // point. The index of each waypoint in the array will be used to mark order restrictions.
+  private List<Point> waypoints;
+  // Array of encoded string representations of the keys of orders from Datastore assigned to the journey.
+  private List<String> orderKeys;
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java:42]: I am a bit confused as to how this will work. Don't you need to store the orders for each waypoint?

Otherwise, how will the courier know what to pick up from each library and what to deliver at each destination? A user might have multiple open orders assigned to multiple couriers, no?

File: src/main/java/com/google/sps/data/JourneyHandler.java [hunk 8]
```diff
@@ -26,19 +27,29 @@
...
+    OrderHandler orderHandler = new OrderHandler(pathFinder);
+    orderHandler.updateStatusForOrders(journey.getOrders(), OrderHandler.OrderStatus.ASSIGNED.toString());
+    addDeliveryJourney(journey, deliverySlot);
   }
 
   /**
    * Adds the details of the delivery journey assigned to deliverySlot to datastore.
    */
-  protected void addDeliveryJourney(ArrayList<Point> waypoints, DeliverySlot deliverySlot) {
+  protected void addDeliveryJourney(Journey journey, DeliverySlot deliverySlot) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/JourneyHandler.java:45]: protected should rarely be used over private. Only when you want to allow inheritance to overwrite it.

File: src/main/java/com/google/sps/data/JourneyHandler.java [hunk 9]
```diff
@@ -26,19 +27,29 @@
...
+    this.pathFinder = pathFinder;
+  }
   /**
    * Finds a set of orders and an optimal journey for deliverySlot.
    */
-  public void processDeliveryRequest(DeliverySlot deliverySlot) throws ApiException, IOException, InterruptedException {
-    DeliverySystem deliverySystem = new DeliverySystem();
-    ArrayList<Point> orderedWaypoints = deliverySystem.findOptimalDeliveryJourneyForDeliverySlot(deliverySlot);
-    addDeliveryJourney(orderedWaypoints, deliverySlot);
+  public void processDeliveryRequest(DeliverySlot deliverySlot) throws ApiException, BadRequestException, DataNotFoundException, EntityNotFoundException, IOException, InterruptedException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/JourneyHandler.java:38]: This should have tests.

I guess the test will return unimplemented for now and you can adjust the test in a future PR where you `implementaddDeliveryJourney`.

In general in the future, try to avoid partly implementing something (e.g. this method does 4 things and then returns unimplemented). Either you defer implementation completely, or you implement something completely.

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 10]
```diff
@@ -152,16 +154,40 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
...
-    return datastore.prepare(orderQuery).asList(FetchOptions.Builder.withDefaults());
+    List<Entity> results = datastore.prepare(orderQuery).asList(FetchOptions.Builder.withDefaults());
+    List <String> resultsKeyStrings = new ArrayList<>();
+    for (Entity order : results) {
+      resultsKeyStrings.add(KeyFactory.keyToString(order.getKey()));
+    }
+    return resultsKeyStrings;
+  }
+
+  /**
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java:173]: New public methods call for new tests in OrderHandlerTest.java

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 11]
```diff
@@ -152,16 +154,40 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
   }
 
   /**
-   * Get unassigned orders from datastore with the area set to area.
-   * TODO[ak47na]: change method to return a list of Order objects so that classes that call it 
-   * don't need to know how data is stored.
+   * Returns the keyString of unassigned orders from datastore with the area set to area.
    */
-  List<Entity> getAvailableOrders(int area) {
+  List<String> getAvailableOrders(int area) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java:159]: You changed this method (good), but no tests needed changing (bad?).

I'm guessing this method is not tested?

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 12]
```diff
@@ -152,16 +154,40 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
...
+   */
+  public Object getProperty(String orderKeyStr, String propertyName) throws EntityNotFoundException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity order = datastore.get(KeyFactory.stringToKey(orderKeyStr));
+    return order.getProperty(propertyName);
+  }
+
+  /** 
+   * Updates the status property of orders with the keyString from orderKeys list to status.
+   TODO
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java]: TODO?

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (alexcepoi) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: Add a TODO somewhere.

Comment (alexcepoi) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: Do you? I see this in Journey:
  private List<Point> waypoints;
  private List<String> orderKeys;


But I don't see how you can infer from that which orders belong to which waypoint?

File: src/test/java/com/google/sps/data/DeliverySystemTest.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      pathFinder.distance(points.get(1), points.get(2)) +
+      pathFinder.distance(points.get(2), points.get(3)) + 
+      pathFinder.distance(points.get(3), points.get(4)) +
+      pathFinder.distance(new Point(-0.0001, 0), points.get(0)) + 1;
+
+    DeliverySlot deliverySlot = new DeliverySlot(new Date(2020, 9, 26), 0, totalSeconds * 1000, "user0");
+    // Start all delivery journey at point (-0.0001, 0).
+    deliverySlot.setStartPoint(-0.0001, 0);
+    Journey journey = deliverySystem.createJourneyForDeliveryRequest(deliverySlot);
+    assertEquals(5, journey.getNumberOfWaypoints());
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/DeliverySystemTest.java:97]: I think you'll want to test here what are the actual waypoints and what are the orders assigned to the waypoints.

Just that there are 5 waypoints might not be enough and doesn't give you enough confidence on what the algorithm does.

File: src/test/java/com/google/sps/data/DeliverySystemTest.java [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    points.add(new Point(0.0005, 0.0006));
+    points.add(new Point(0.0009, 0.0008));
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test(expected = UnsupportedOperationException.class)
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/DeliverySystemTest.java:82]: Oh wow.. this confused me so much :)))

You actually wrote the entire test, but none of it passes and returns unimplemented in the first line. This is somewhat uncommon, and you should look at ways you can avoid it in the future, as you kind of similar to just submitting comented out code (which is an antipattern).

The split between this PR and #31 is probably not that successful, as there's not a clear boundary as to what this PR does on what does #31 do. I don't have great ideas on how this could be improved, and no point in undoing it now though.

**Review by alexcepoi — APPROVED** [review]
Approved, but take a look at the above comments about fixing your test coverage before merging.

File: src/main/java/com/google/sps/data/JourneyHandler.java (hunk 8 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/JourneyHandler.java:45]: I changed `addDeliveryJourney` to private for now, but if I want to test the method, is it okay to make it public?

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: Done.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 10 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java:173]: I added one test for it.

File: src/main/java/com/google/sps/data/JourneyHandler.java (hunk 8 above)
Comment (alexcepoi) [reply src/main/java/com/google/sps/data/JourneyHandler.java:45]: In general, it's a bad idea to test implementation details, and private means something is an implementation detail. It's generally recommended to test all your public methods, since public methods are what the class advertises to external callers that they can and should call.

In your case processDeliveryRequest doesn't really do much other than call into other well tested methods. So I'd actually consider removing processDeliveryRequest altogether and just keeping a public `addDeliveryJourney`.

You might consider calling `orderHandler.updateStatusForOrders` inside `addDeliveryJourney`, as it doesn't really ever make sense to do these 2 actions separately (they should actually eventually happen within the same transaction, and having a transaction spawning multiple methods/classes makes things actually quite complicated).

File: src/test/java/com/google/sps/data/DeliverySystemTest.java (hunk 13 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/DeliverySystemTest.java:97]: Yes, I will add tests for the actual points in https://github.com/googleinterns/step235-2020/pull/31.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 12 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java]: I changed the comment.

File: src/main/java/com/google/sps/data/JourneyHandler.java (hunk 9 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/JourneyHandler.java:38]: I changed the method to only throw `UnsupportedOperationException`.

File: src/main/java/com/google/sps/data/DeliverySystem.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/DeliverySystem.java]: I added the orders to points here: https://github.com/googleinterns/step235-2020/pull/33.
